src/main/java/fi/pohina/vinkkilista/domain/BookmarkService.java

```diff
@@ -1,43 +1,209 @@
         return UUID.randomUUID().toString();
     }
+
+    /**
+     * Checks and fixes the given tag set.
+     *
+     * @param tags Tags in a string set
+     * @return Tags in a string set
+     */
+    public Set<String> validateTagSet(Set<String> tags) {
+        Set<String> tagsSet = new HashSet<>();
+
+        for (String tag : tags) {
+            tag = validateTag(tag);
+
+            if (tag != null) {
+                tagsSet.add(tag);
+            }
+        }
+
+        return tagsSet;
+    }
+
+    /***
+     * Function for validating a tag.
+     * @param tag
+     * @return A validated tag as a string that contains no extra spaces and 
+     * has only allowed characters. Null if length is 0.
+     * Currently Only allows alpha-numeric characters.
+     */
+    public String validateTag(String tag) {
+
+        tag = tag.toLowerCase();
+
+        StringBuilder validated = new StringBuilder();
+
+        for (int i = 0; i < tag.length(); i++) {
+            char c = tag.charAt(i);
+
+            if (allowedCharacter(c)) {
+                validated.append(c);
+            }
+        }
+
+        tag = validated.toString().trim().replaceAll(" +", " ");
+
+        if (tag.length() == 0) {
+            return null;
+        }
+
+        return tag;
+    }
+
+    private boolean allowedCharacter(char c) {
+
+        if (c >= 'a' && c <= 'z') {
+            return true;
+        }
+
+        if ((c >= '0' && c <= '9') || c == ' ') {
+            return true;
+        }
+        
+        return false;
+    }
+    
+    /**
+     * Function for converting string tag set to a set of existing tag objects,
+     * and alternatively also creates missing tags if specified. Validates
+     * the given set of tags.
+     *
+     * @param tags comma-separated string list of tags
+     * @param createNew Boolean for whether to create missing tags
+     * @return set of tag objects
+     */
+    public Set<Tag> tagSetStringToObject(Set<String> tags, Boolean createNew) {
+        Set<Tag> tagsSet = new HashSet<>();
+
+        tags = validateTagSet(tags);
+
+        for (String tagString : tags) {
+
+            Tag tagObject = tagDao.findByName(tagString);
+            if (tagObject != null && !tagsSet.contains(tagObject)) {
+                tagsSet.add(tagObject);
+            } else if (createNew) {
+                tagsSet.add(createTag(tagString));
+            }
+        }
+        
+        return tagsSet;
+    }
+
+    /**
+     * Takes the URL given and returns the appropriate tag 
+     * related to that URL as a string.
+     *
+     * @param url The URL of the bookmark
+     * @return Name of the tag as string, empty if no match
+     */
+    public String addTagStringByUrl(String url) {
+        String[] videoUrls = {"youtube.com", "vimeo.com", "youtu.be"};
+        String[] blogUrls = {
+            "blogger.com", "blogs.helsinki.fi", 
+            "wordpress.org", "blogspot.com"};
+        String[] bookUrls = {".suomalainen.com"};
```

Comment:
Why is suomalainen with a . but the others are not?

Reply:
Because just suomalainen.com is more likely to occur on other sites than the intended one, in addition if this were to happen it is unlikely something with a suomalainen.com url has something to do with books.